ReactNativeClient/lib/services/queryBuilder.ts

```diff
@@ -0,0 +1,71 @@
+interface Term {
+  relation: string;
+  value: string;
+}
+
+// Note: we might need to escape query for making it sql compatible
+const matchStmtBuilder = (
+	filters: Map<string, Array<Term>>,
+	keyword: string
+): string => {
+	if (filters.has(keyword)) {
+		return `${filters
+			.get(keyword)
+			.map((term) => {
+				if (term.relation === 'AND') {
+					if (keyword !== 'text') {
+						return `${keyword}:${term.value}`;
+					} else {
+						return `${term.value}`;
+					}
+				} else {
+					return `${term.relation} ${keyword}:${term.value}`;
+				}
+			})
+			.join(' ')} `;
+	}
+	return '';
+};
+
+
+module.exports = (filters: Map<string, Array<Term>>) => {
+	// console.log(filters);
+	let match = '';
+	match += matchStmtBuilder(filters, 'title');
+	match += matchStmtBuilder(filters, 'body');
+	match += matchStmtBuilder(filters, 'text');
+
+	let query = `SELECT
+	notes_fts.id,
+	notes_fts.title AS normalized_title,
+	offsets(notes_fts) AS offsets,
+	notes.title,
+	notes.user_updated_time,
+	notes.is_todo,
+	notes.todo_completed,
+	notes.parent_id
+	FROM notes_fts
+	LEFT JOIN notes ON notes_fts.id = notes.id WHERE 1`;
+
+	const params = [];
+
+	if (match !== '') {
+		// there is something to fts search
+		query += ' AND notes_fts MATCH (?)';
+		params.push(match.trim());
+	}
+
+	// create a tag filter
+	if (filters.has('tag')) {
+		query += ' AND notes.id IN (SELECT note_tags.note_id FROM note_tags WHERE 1';
+
+		filters.get('tag').forEach((term) => {
+			if (term.relation === 'AND') query += ' INTERSECT SELECT note_tags.note_id FROM note_tags WHERE note_tags.tag_id IN (select tags.id from tags WHERE tags.title LIKE ?)';
+			else if (term.relation === 'OR') query += ' UNION SELECT note_tags.note_id FROM note_tags WHERE note_tags.tag_id IN (select tags.id from tags WHERE tags.title LIKE ?)';
+			params.push(term.value);
+		});
+		query += ')';
+	}
```

Comment:
Please be careful here as you've replaced the MATCH query with a regular query, which is going to be significantly slower. In fact, I think there's no point using FTS if we aren't going to use MATCH queries. See the example in the doc: https://sqlite.org/fts3.html A regular query takes 22.5 seconds and a match one takes 0.03 second. So please check if you can write the query using MATCH.

Reply:
I already read the docs and understand your concern. 

But the problem does not exist here. At line 54 we already do the expensive search using a MATCH statement. The regular queries are all just to filter the note ids returned by the MATCH. These are inexpensive queries of the small tags and note_tags table (which also aren't fts and so we do need regular queries).

Reply:
The MATCH query is not replaced with a regular query. We are just adding an additional constraint to the output of the MATCH.

I tested using 8000 notes and it's still just as fast as before.

Reply:
The MATCH query is not replaced with a regular query. We're just adding an additional constraint to the output of the MATCH. I tested using 8000 notes and it's still just as fast as before.

Admittedly we don't really need the join if there is no text to FTS search but it is not a performance bottleneck and is still quite fast. I'll do additional testing.